File: Shadow/JBM0002JT_KengenApp/app/src/main/java/com/water/alkaline/kengen/ui/fragment/HistoryFragment.java
package com.water.alkaline.kengen.ui.fragment;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;

import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.preference.PowerPreference;
import com.water.alkaline.kengen.R;
import com.water.alkaline.kengen.databinding.FragmentHistoryBinding;
import com.water.alkaline.kengen.library.ItemOffsetDecoration;
import com.water.alkaline.kengen.model.feedback.Feedback;
import com.water.alkaline.kengen.ui.activity.FeedbackActivity;
import com.water.alkaline.kengen.ui.adapter.FeedAdapter;
import com.water.alkaline.kengen.utils.Constant;



import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link HistoryFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class HistoryFragment extends Fragment {


    FragmentHistoryBinding binding;
    Activity activity;
    List<Feedback> feedbacks = new ArrayList<>();
    FeedAdapter adapter;

    public HistoryFragment(Activity activity) {
        this.activity = activity;
    }


    public HistoryFragment() {
    }

    public static HistoryFragment newInstance(Activity activity) {
        return new HistoryFragment(activity);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (activity == null) {
            activity = (FeedbackActivity) context;
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentHistoryBinding.inflate(getLayoutInflater(), container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (activity != null) {
            GridLayoutManager manager = new GridLayoutManager(activity, 1);
            manager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int i) {
                    switch (adapter.getItemViewType(i)) {
                        case Constant.STORE_TYPE:
                            return 1;
                        case Constant.AD_TYPE:
                            return 1;
                        case Constant.LOADING:
                            return 1;
                        default:
                            return 1;

                    }
                }
            });
            binding.rvFeeds.setLayoutManager(manager);
            adapter = new FeedAdapter(activity, feedbacks);
            binding.rvFeeds.setAdapter(adapter);
            binding.rvFeeds.setItemViewCacheSize(100);

            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    refreshActivity();
                }
            }, 500);
        }
    }

    public void refreshActivity() {
        try {
            Type type = new TypeToken<List<Feedback>>() {
            }.getType();

            feedbacks = new Gson().fromJson(PowerPreference.getDefaultFile().getString(Constant.mFeeds, new Gson().toJson(new ArrayList<Feedback>())), type);
        } catch (Exception e) {
            feedbacks = new ArrayList<>();
        }
        adapter.refresh(feedbacks);
        binding.includedProgress.progress.setVisibility(View.GONE);
        checkData();
    }

    public void checkData() {
        if (binding.rvFeeds.getAdapter() != null && binding.rvFeeds.getAdapter().getItemCount() > 0) {
            binding.includedProgress.llError.setVisibility(View.GONE);
        } else {
            binding.includedProgress.llError.setVisibility(View.VISIBLE);
        }
    }

}

File: KengenWaterApp/KengenWaterApp/app/src/main/java/com/water/alkaline/kengen/ui/feedback/fragment/HistoryFragment.kt
package com.water.alkaline.kengen.ui.feedback.fragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import com.water.alkaline.kengen.databinding.FragmentHistoryBinding
import com.water.alkaline.kengen.model.feedback.Feedback
import com.water.alkaline.kengen.ui.base.BaseFragment
import com.water.alkaline.kengen.ui.feedback.FeedbackActivity
import com.water.alkaline.kengen.utils.Constant
import com.water.alkaline.kengen.utils.FeedBacksEvent
import com.water.alkaline.kengen.utils.NewFeedBackEvent
import dagger.hilt.android.AndroidEntryPoint
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode


@AndroidEntryPoint
class HistoryFragment : BaseFragment() {

    private val binding by lazy {
        FragmentHistoryBinding.inflate(layoutInflater)
    }
    private var adapter: FeedAdapter? = null


    override fun onAttach(context: Context) {
        super.onAttach(context)
        context.setMainContext()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setAdapter()
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onNewFeedBackEvent(event: NewFeedBackEvent) {
        adapter!!.addItem(event.model)
        checkData()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onFeedBacksEvent(event: FeedBacksEvent) {
        refreshActivity(event.feedbacks)
    }


    private fun setAdapter() {
        val manager = GridLayoutManager(appContext, 1)
        manager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(i: Int): Int {
                return when (adapter!!.getItemViewType(i)) {
                    Constant.STORE_TYPE -> 1
                    Constant.AD_TYPE -> 1
                    Constant.LOADING -> 1
                    else -> 1

                }
            }
        }
        binding.rvFeeds.layoutManager = manager
        adapter = FeedAdapter(
            appContext,
            mutableListOf()
        )
        binding.rvFeeds.adapter = adapter
        binding.rvFeeds.setItemViewCacheSize(100)
    }

    private fun refreshActivity(feedbacks: List<Feedback>) {
        adapter!!.refresh(feedbacks)
        binding.includedProgress.progress.visibility = View.GONE
        checkData()
    }

    @Subscribe

    private fun checkData() {
        if (adapter!!.itemCount > 0) {
            binding.includedProgress.llError.visibility = View.GONE
        } else {
            binding.includedProgress.llError.visibility = View.VISIBLE
        }
    }
}